### ISSUE TITLE
Custom namespaces for custom exceptions?

### ISSUE THREAD

**Issue by panosru** [omegad-biz/fault-manager#4]
I'm still thinking about that.

Currently, all generated exceptions are in the global namespace.

So if you run `Fault::exception('MyCustomException');` it will result in `\MyCustomException`.

Is not a hard implementation, but I'm yet to come up if is a good idea.

Also what you prefer?

1. `Fault::exception('\CustomNamespace\MyCustomException');`
2. `Fault::exception('CustomNamespace\MyCustomException');` (without the leading backslash)
3. `Fault::exception('MyCustomException', <message>, <code>, <previous>, <arguments>, 'CustomNamespace');`
  3.1 `Fault::exception('MyCustomException', null, null, null, null, 'CustomNamespace');`

I don't like the 3rd way 😵 

@alextech any thoughts?

**Comment by alextech** [comment]
I will stand by my original \CustomException::class
It is used in style guide, public OSS code reviews make you change it, and there were full projects across libraries converting arbitrary strings to this standard not action.

**Comment by panosru** [comment]
So basically you are going with option 2 correct?

having either `CustomNamespace\MyCustomException` or `\CustomNamespace\MyCustomException::class` is the same thing.

```php
var_dump(
    'CustomNamespace\MyCustomException',
    \CustomNamespace\MyCustomException::class
);
```
[^ Code Result](https://3v4l.org/NQrK9)

**Comment by alextech** [comment]
Right. Im honestly not sure what exactly :::class generates... I should have checked first. I think it should have slash in front to ensure there is no ambiguity over whether its a relative namespace from where its specified, or whole new one.

**Comment by panosru** [comment]
It generates exactly what you would type in `namespace` keyword.

for instance, you would type that:
```php
namespace Foo;

class Bar {}
```
because the following would produce a fatal error:
```php
namespace \Foo;

class Bar {}
```
[^ Code Result](https://3v4l.org/Z8k3m)

Thus the `::class` outputs the namespace without the leading backslash as shown in [php manual](http://php.net/manual/en/language.oop5.basic.php#language.oop5.basic.class.class)

**Comment by alextech** [comment]
Oh interesting. Good to know (if I can remember). Thanks for info.

**Comment by panosru** [comment]
So, the main question is yet to be answered, do we need custom namespaces for that matter?

Any use cases?

**Comment by alextech** [comment]
Cannot think of any. If I am inside of \Importer, I am not going to add \importerexception. I will continue using Importer\InvalidRowException
Or did I misunderstand again, and that *is* custom namespace. My interpretation of custom is something that doesn't exist at all. Or you mean anything other than root?

**Comment by panosru** [comment]
Let's say that you have a library called `importer` that works under namespace `\Importer` and has it's own exceptions already defined, for instance, `\Importer\InvalidRowException`.

When you want to throw your extension you would do:
```php
throw new \Importer\InvalidRowException('Custom message', 123);
```

With `FaultManager` you can have two more ways to throw your already defined exception:
```php
//First way
throw Fault::exception(\Importer\InvalidRowException::class, 'Custom message', 123);

//Second way
Fault::throw(\Importer\InvalidRowException::class, 'Custom message', 123);
```
That doesn't mean that you have to do that but is there so you can do it either way.

The point with `FaultManager` is to not end up with a bunch of exceptions that simply extend `\Exception` and nothing more.
```php
class SimpleException extends \Exception
{
}
```
So you can simply use:
```php
throw Fault::exception(\SimpleException::class, 'my message');
// or
Fault::throw(\SimpleException::class, 'my message');
```
and a class file by name `_compiled/SimpleException.php` under predefined compile path with the contents:

```php
class SimpleException extends \Exception
{
}
```

So the question is, would it make sense to generate custom exceptions under a namespace so you would have:
```php
Fault::throw(\Importer\SimpleException::class, 'my message');
```
And a file will be generated under the predefined path `_compiled/Importer/SimpleException.php` 

or `_compiled/Importer_SimpleException.php`?
